Forbidden float column in short key

EmmyMiao87 opened the pull request:
When the user does not specify the key column, doris will automatically supplement the key column.
However, doris does not support float or double as the key column, so when adding the key column, doris should avoid setting those column as the key column.

The CreateMaterailizedView, AddRollup and CreateDuplicateTable need to forbidden float column in short key.
If the float column is directly encountered during the supplement process, the subsequent columns are all value columns.
If the first column is float or double, Doris will throw the exception.

Fixed #3811

Change-Id: Ib66d9355acefcd8f281906bcb7b4dd684319ff08

EmmyMiao87 committed:
Forbidden float column in short key

When the user does not specify the key column, doris will automatically supplement the key column.
However, doris does not support float or double as the key column, so when adding the key column, doris should avoid setting those column as the key column.

The CreateMaterailizedView, AddRollup and CreateDuplicateTable need to forbidden float column in short key.
If the float column is directly encountered during the supplement process, the subsequent columns are all value columns.
If the first column is float or double, Doris will throw the exception.

Fixed #3811

Change-Id: Ib66d9355acefcd8f281906bcb7b4dd684319ff08

chaoyli commented:
Create table will forbidden float/double

morningman commented on fe/src/main/java/org/apache/doris/analysis/CreateMaterializedViewStmt.java:
```diff
@@ -235,6 +236,9 @@ private void analyzeOrderByClause() throws AnalysisException {
                         break;
                     }
+                    if (mvColumnItem.getType().isFloatingPointType()) {
+                        throw new AnalysisException("Float or double can not used as a key, use decimal instead.");
```
```suggestion
                        throw new AnalysisException("Float or double can not used as a sort key, use decimal instead.");
```

morningman commented on fe/src/main/java/org/apache/doris/analysis/CreateMaterializedViewStmt.java:
```diff
@@ -255,10 +259,15 @@ private void analyzeOrderByClause() throws AnalysisException {
                 keyStorageLayoutBytes += resultColumn.getType().getStorageLayoutBytes();
-                if ((i + 1) <= FeConstants.shortkey_max_column_count
-                        || keyStorageLayoutBytes < FeConstants.shortkey_maxsize_bytes) {
+                if ((!mvColumnItem.getType().isFloatingPointType())
```
We should be break if we met a floating point type column. The following case will get wrong result in your logic:

`k1 int, k2 float, k3 int`

EmmyMiao87 replied:
Why?

EmmyMiao87 commented:
> Create table will forbidden float/double

Already forbidden

EmmyMiao87 committed:
Update fe/src/main/java/org/apache/doris/analysis/CreateMaterializedViewStmt.java

Co-authored-by: Mingyu Chen <morningman.cmy@gmail.com>

EmmyMiao87 committed:
Fix bug

Change-Id: Id5236ff7229fc1c1d35df579dd3c62e699392ba5

EmmyMiao87 committed:
Change the error msg

Change-Id: Ie1320d9af56083633c7634c1c3df485f3994da31

EmmyMiao87 committed:
Change code style

Change-Id: Ib4c4791847d522fa312efff4016787c9caeab785

EmmyMiao87 committed:
Change the strategy of order by column(key column)

For duplicate table without order by columns, the order by columns are same as short key columns.
If the order by columns have been designated, the count of short key columns must be <= the count of order by columns.
The short key columns must be less then 3 columns and less then 36 bytes.
Also the float and double could not be the short key column. At the same time, doirs must be at least one short key column.
So the type of first column could not be float or double.
If the varchar is the short key column, it can only be the least one short key column.

Change-Id: I46ec2c78e87d73c3f3aeeb815b9acc6dcfe280e0

EmmyMiao87 committed:
Remove unused

Change-Id: Ia4d2a9bfb1b4321dc3ed8e2dd243d9a4fd2dbf69

EmmyMiao87 committed:
Change comment

Change-Id: I1d13600a39d59cf93feb42df85c1089ca1029615

EmmyMiao87 committed:
Remove unused import

Change-Id: Id647f940b81ebfc27e18ef83e5275af7177bf926

EmmyMiao87 committed:
Fix bug

Change-Id: I4d560427156855f28598ed089048006ea8d685c4

EmmyMiao87 committed:
Fix bug

Change-Id: Iccfce893d120506667c55444a11841b91a1d322b

EmmyMiao87 committed:
Fix bug

Change-Id: Ia88fa8cfa70661328300864054962931a8c0e225

EmmyMiao87 committed:
Fix ut

Change-Id: I88e36a2379ae2e0659358bb5ba59d93bccfe37f1

morningman reviewed (dismissed):
LGTM

morningman reviewed (approved):
LGTM
